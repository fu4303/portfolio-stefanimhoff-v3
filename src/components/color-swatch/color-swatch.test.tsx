import React from 'react';
import { axe } from 'jest-axe';
import user from '@testing-library/user-event';

import { render, fireEvent } from '../../services/test-utils';

import { ColorSwatch } from '.';
import type { ColorSwatchProps } from '.';

type ColorSwatchTestProps = ColorSwatchProps & {
  title: string;
};

const defaultProps = {
  color: '#005CAF',
  name: 'Lapis Lazuli',
  description:
    'The soft, slightly purplish blue associated with the semi-precious stone.',
};

describe('ColorSwatch', () => {
  // prettier-ignore
  const combinations: ColorSwatchTestProps[] = [
    { title: 'default', color: '#005CAF', name: 'name', description: 'description' },
    { title: 'description', color: '#005CAF', description: 'description' },
    { title: 'rgb', color: 'rgb(0, 92, 175)' }, 
    { title: 'hsl', color: 'hsl(208, 100%, 34%)' },
  ];

  combinations.forEach((combination) => {
    test(`renders in ${combination.description}`, () => {
      const { container } = render(<ColorSwatch {...combination} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  combinations.forEach((combination) => {
    test(`is accessible in ${combination.description}`, async () => {
      const { container } = render(<ColorSwatch {...combination} />);
      const results = await axe(container);

      expect(results).toHaveNoViolations();
    });
  });

  test('renders correctly with name', () => {
    const { getByText } = render(<ColorSwatch {...defaultProps} />);
    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });

  test('renders correctly with description', () => {
    const { getByText } = render(
      <ColorSwatch {...defaultProps} name={undefined} />
    );
    expect(getByText(/purplish blue/i)).toBeInTheDocument();
  });

  test('renders correctly with RGB value', () => {
    const { getByText } = render(<ColorSwatch color="rgb(0, 92, 175)" />);
    expect(getByText(/#005CA/)).toBeInTheDocument();
  });

  test('renders correctly with HSL value', () => {
    const { getByText } = render(<ColorSwatch color="hsl(208, 100%, 34%)" />);
    expect(getByText(/#005CA/)).toBeInTheDocument();
  });

  test('show color information on click', async () => {
    const { getByText, findByLabelText } = render(
      <ColorSwatch {...defaultProps} />
    );

    user.click(await findByLabelText('Information'));
    expect(getByText(/#005CA/i)).toBeInTheDocument();

    user.click(await findByLabelText('Information'));
    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });

  test('should cycle through information icon', async () => {
    const { findByLabelText } = render(<ColorSwatch {...defaultProps} />);

    expect(document.body).toHaveFocus();
    user.tab();
    expect(await findByLabelText('Information')).toHaveFocus();
  });

  test('show color information on Enter keyUp', async () => {
    const { getByText, findByLabelText } = render(
      <ColorSwatch {...defaultProps} />
    );

    fireEvent.keyUp(await findByLabelText('Information'), {
      key: 'Enter',
      code: 13,
    });
    expect(getByText(/#005CA/i)).toBeInTheDocument();

    fireEvent.keyUp(await findByLabelText('Information'), {
      key: 'Enter',
      code: 13,
    });
    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });
});
