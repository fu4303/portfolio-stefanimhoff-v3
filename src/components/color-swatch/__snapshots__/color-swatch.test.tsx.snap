// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSwatch renders in description 1`] = `
.c0 {
  border-color: var(--colorForegroundSoft);
  border-radius: 5px;
  border-style: solid;
  border-width: 1px;
  box-shadow: 0 0 50px rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 0.8rem;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
}

.c0:hover,
.c0:focus {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c1 {
  background: lightgrey;
  border-bottom-color: var(--colorForegroundSoft);
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  min-height: 6rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  right: 0;
  top: 0;
  width: 30px;
}

.c4 {
  fill: var(--colorForegroundFullContrast);
  font-size: 1rem;
}

.c2 {
  background: var(--colorBackgroundFullContrast);
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  color: var(--colorForegroundFullContrast);
  height: 100%;
  min-height: 8rem;
  padding: 0.8rem 0.6rem;
  position: relative;
}

.c5 {
  font-size: 0.8rem;
  font-weight: bold;
  line-height: 1;
  padding-bottom: 0.3rem;
  padding-right: 20px;
}

.c6 {
  margin-bottom: 0.5rem;
  padding-right: 20px;
}

<div
  class="c0"
  title="default"
>
  <div
    class="c1"
    style="background-color: rgb(0, 92, 175);"
  />
  <div
    class="c2"
  >
    <div
      aria-label="Information"
      class="c3"
      role="button"
      tabindex="0"
    >
      <svg
        aria-label="Information Icon"
        class="c4"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 100-16 8 8 0 000 16zM11 7h2v2h-2V7zm0 4h2v6h-2v-6z"
        />
      </svg>
    </div>
    <div
      class="c5"
    >
      name
    </div>
    <div
      class="c6"
    >
      description
    </div>
  </div>
</div>
`;

exports[`ColorSwatch renders in description 2`] = `
.c0 {
  border-color: var(--colorForegroundSoft);
  border-radius: 5px;
  border-style: solid;
  border-width: 1px;
  box-shadow: 0 0 50px rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 0.8rem;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
}

.c0:hover,
.c0:focus {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c1 {
  background: lightgrey;
  border-bottom-color: var(--colorForegroundSoft);
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  min-height: 6rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  right: 0;
  top: 0;
  width: 30px;
}

.c4 {
  fill: var(--colorForegroundFullContrast);
  font-size: 1rem;
}

.c2 {
  background: var(--colorBackgroundFullContrast);
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  color: var(--colorForegroundFullContrast);
  height: 100%;
  min-height: 8rem;
  padding: 0.8rem 0.6rem;
  position: relative;
}

.c5 {
  font-size: 0.8rem;
  font-weight: bold;
  line-height: 1;
  padding-bottom: 0.3rem;
  padding-right: 20px;
}

.c6 {
  margin-bottom: 0.5rem;
  padding-right: 20px;
}

<div
  class="c0"
  title="description"
>
  <div
    class="c1"
    style="background-color: rgb(0, 92, 175);"
  />
  <div
    class="c2"
  >
    <div
      aria-label="Information"
      class="c3"
      role="button"
      tabindex="0"
    >
      <svg
        aria-label="Information Icon"
        class="c4"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 100-16 8 8 0 000 16zM11 7h2v2h-2V7zm0 4h2v6h-2v-6z"
        />
      </svg>
    </div>
    <div
      class="c5"
    />
    <div
      class="c6"
    >
      description
    </div>
  </div>
</div>
`;

exports[`ColorSwatch renders in undefined 1`] = `
.c0 {
  border-color: var(--colorForegroundSoft);
  border-radius: 5px;
  border-style: solid;
  border-width: 1px;
  box-shadow: 0 0 50px rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 0.8rem;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
}

.c0:hover,
.c0:focus {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c1 {
  background: lightgrey;
  border-bottom-color: var(--colorForegroundSoft);
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  min-height: 6rem;
}

.c2 {
  background: var(--colorBackgroundFullContrast);
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  color: var(--colorForegroundFullContrast);
  height: 100%;
  min-height: 8rem;
  padding: 0.8rem 0.6rem;
  position: relative;
}

.c3 {
  font-size: 0.8rem;
}

.c4 {
  padding-bottom: 0.1rem;
  padding-top: 0.1rem;
}

<div
  class="c0"
  title="rgb"
>
  <div
    class="c1"
    style="background-color: rgb(0, 92, 175);"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <strong>
          R
        </strong>
         
        0
         
        <strong>
          G
        </strong>
         
        92
         
        <strong>
          B
        </strong>
         
        175
         
      </div>
      <div
        class="c4"
      >
        <strong>
          H
        </strong>
         
        208
         
        <strong>
          S
        </strong>
         
        100
         
        <strong>
          L
        </strong>
         
        34
         
      </div>
      <div
        class="c4"
      >
        <strong>
          #005CAF
        </strong>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorSwatch renders in undefined 2`] = `
.c0 {
  border-color: var(--colorForegroundSoft);
  border-radius: 5px;
  border-style: solid;
  border-width: 1px;
  box-shadow: 0 0 50px rgba(0,0,0,0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 0.8rem;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
}

.c0:hover,
.c0:focus {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c1 {
  background: lightgrey;
  border-bottom-color: var(--colorForegroundSoft);
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  min-height: 6rem;
}

.c2 {
  background: var(--colorBackgroundFullContrast);
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  color: var(--colorForegroundFullContrast);
  height: 100%;
  min-height: 8rem;
  padding: 0.8rem 0.6rem;
  position: relative;
}

.c3 {
  font-size: 0.8rem;
}

.c4 {
  padding-bottom: 0.1rem;
  padding-top: 0.1rem;
}

<div
  class="c0"
  title="hsl"
>
  <div
    class="c1"
    style="background-color: rgb(0, 92, 173);"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <strong>
          R
        </strong>
         
        0
         
        <strong>
          G
        </strong>
         
        92
         
        <strong>
          B
        </strong>
         
        173
         
      </div>
      <div
        class="c4"
      >
        <strong>
          H
        </strong>
         
        208
         
        <strong>
          S
        </strong>
         
        100
         
        <strong>
          L
        </strong>
         
        34
         
      </div>
      <div
        class="c4"
      >
        <strong>
          #005CAD
        </strong>
      </div>
    </div>
  </div>
</div>
`;
